### ISSUE TITLE
Sending email in LiveView process breaks tests

### ISSUE THREAD

**Issue by jc00ke** [swoosh/swoosh#430]
I just upgraded to 0.24.3 and I'm now seeing this error when I run this tests with other tests with `mix test`. If I run it by itself, or just the tests in that one file, it passes.

```ex
test "sends email on confirmation", %{conn: conn, participant: participant, session: session} do
  {:ok, view, _html} = live_isolated(conn, View, session: session)
  render_click(view, :confirming)
  assert render_click(view, :confirm, %{participant_id: participant.id}) =~ "Sending..."
  assert render(view) =~ "Sent!"
  assert_email_sent(to: participant.user.email)
end
```

```
** (FunctionClauseError) no function clause matching in FooWeb.ButtonLive.handle_info/2
    (foo) lib/foo_web/views/live/button_live.ex:56: FooWeb.ButtonLive.handle_info({:email, %Swoosh.Email{assigns: %{...}}...
```

```ex
def handle_info({:send_email, participant_id}, socket) do
  participant_id
  |> FooWeb.Participant.ChangeInfoEmail.generate()
  |> FooWeb.Mailer.deliver()

  {:noreply, socket}
end
```

I'm guessing the `pids/0` [function](https://github.com/swoosh/swoosh/commit/15b57f9df4443280016e5273ad96cb448a3a86c0#diff-8a2aba08d459118194bfeb9749c5e83aR30) is including the `LiveView` process when it should only include the test process?

**Comment by princemaple** [comment]
Hi @jc00ke is `view` a process? The test would never have worked, would it? Before the linked change, TestAdapter only sent message to the process that called `deliver`. And the assertion would only have worked if `deliver` was called in the test process. After the change, the message can be received by the test process, where the assertion is called, but admittedly it's going to cause trouble in scenarios like this. The current implementation is probably not very suitable for integration test.

Please feel free to correct me if I'm wrong. I'd surprised if the test used to work.

Regarding how to properly do it for integration tests, it's going to take some thinking...

**Comment by jc00ke** [comment]
@princemaple previously I was using a custom adapter:

```ex
defmodule FooWeb.Mailer.TestAdapter do
  use Swoosh.Adapter

  def deliver(email, _config) do
    send(test_process(), {:email, email})
    {:ok, %{}}
  end

  defp test_process do
    Application.get_env(:swoosh, :shared_test_process) || self()
  end
end
```
and in my tests, where necessary, I would do this:

```ex
Application.put_env(:swoosh, :shared_test_process, self())
```

**Comment by princemaple** [comment]
That makes sense. I'll take another look soon.

**Comment by princemaple** [comment]
Hi @jc00ke are you no longer using your custom TestAdapter? As far as I can tell, if not running integration tests in parallel, your custom adapter should just work.

**Comment by jc00ke** [comment]
@princemaple I'm going to add it back for that one test, then I should be gtg until a more permanent solution emerges. Thanks!

**Comment by princemaple** [comment]
Cool. I'll leave this one open and find some time to think of a better solution. I have some very weird ideas. 😉

**Comment by princemaple** [comment]
Oh wow. I'm sorry this caused that much hassle.

**Comment by princemaple** [comment]
My new ideas do require different adapters for unit test and integration test though. Still forming.

**Comment by jc00ke** [comment]
@princemaple no worries! It's not a big deal to use a mock in this case. Not ideal... but what is? 😆

**Comment by princemaple** [comment]
What do you think about the idea that we have a different test adapter that instead of sending a message, blocks and tries to receive one. Blocking will be fine since it's intended to be used in processes. And a different assertion that instead of trying to receive a message, sends one to the process that's trying to send an email (hence blocking and trying to receive the message the assertion sends).

A new problem that will need solving is, how to easily use different adapters for different tests.

**Comment by jc00ke** [comment]
How would that work when tests are run asynchronously? Wouldn't the overall adapter have to be global?

**Comment by princemaple** [comment]
https://github.com/swoosh/swoosh/pull/437 Please take a look at this PR

**Comment by jc00ke** [comment]
@princemaple very cool! How do you choosing the adapter though? Seems like you can't swap it out globally (in `config/test.exs`, for example) w/o potentially affecting other tests. Maybe I'm missing something?

**Comment by princemaple** [comment]
Yeah that's the problem. Still need some consideration. :) The PR is just PoC for now.

**Comment by princemaple** [comment]
As a last resort, we can have integration tests run in a separate batch, like `mix test.integration` and have the adapter set to the AsyncTest.

**Comment by jc00ke** [comment]
I'm wondering if all that ceremony is better than just passing a mock mailer around?

**Comment by princemaple** [comment]
You are not wrong, at this stage a mock is probably as good. After all what is being tested is just that deliver is being called :sweat:

**Comment by princemaple** [comment]
Closing as won't fix, unless we come up with stronger demands and better ideas 👀

**Comment by paulstatezny** [comment]
Bummer this is a `wontfix`. I'm having this problem, and the increasing amount of people using LiveView will too.

**Comment by princemaple** [comment]
https://github.com/swoosh/swoosh/pull/565/files

Give this a go.

**Comment by princemaple** [comment]
https://github.com/swoosh/swoosh/blob/main/CHANGELOG.md#-features-3

**Comment by paulstatezny** [comment]
Thanks @princemaple 😄 Looks like it was fixed, just not commented here. Thanks for all your work on Swoosh ❤️

**Comment by princemaple** [comment]
😃

**Comment by treeman** [comment]
Just ran into this, and `setup :set_swoosh_global` fixed it for me. Thanks!
